File: apps/ios/Anky/Core/Protocol/AnkyDuration.swift
import Foundation

public enum AnkyDuration {
    public static let completeRitualMinutes = 8
    public static let completeRitualMs: Int64 = Int64(completeRitualMinutes) * 60 * 1000

    // Geshtu D3 — silence duration is a setting; the sentinel token is not.
    //
    // The inactivity threshold (how long the writing must rest before the
    // channel closes) is user-configurable: default 8s, bounds 3s–30s. It is
    // routed through this one clamp so no other code re-invents the bounds.
    //
    // The `8000` end-sentinel is the canonical marker a sealed protocol string
    // carries — a symbol, not a measurement. It stays fixed regardless of the
    // configured silence duration, so the parse-time "is this sealed?" check
    // and any historical artifact keep meaning the same thing forever.
    public static let defaultTerminalSilenceMs: Int64 = 8000
    public static let minTerminalSilenceMs: Int64 = 3000
    public static let maxTerminalSilenceMs: Int64 = 30000
    public static let terminalSilenceMs: Int64 = defaultTerminalSilenceMs

    /// The fixed marker written to close a protocol string. Never varies with
    /// the configured inactivity threshold (D3).
    public static let canonicalSentinelToken: Int64 = 8000

    /// Clamp a *configured inactivity threshold* into the allowed 3s–30s band.
    /// This is the single source of the silence-duration bounds.
    public static func clampedTerminalSilenceMs(_ value: Int64) -> Int64 {
        min(max(value, minTerminalSilenceMs), maxTerminalSilenceMs)
    }

    public static func terminalMarkerMs(from line: String) -> Int64? {
        guard !line.isEmpty,
              line.allSatisfy(\.isNumber),
              let value = Int64(line),
              value > 0 else {
            return nil
        }
        return value
    }

    public static func writingDurationMs(_ parsed: ParsedAnky) -> Int64 {
        parsed.events.reduce(Int64(0)) { sum, event in
            sum + event.deltaMs
        }
    }

    public static func durationMs(_ parsed: ParsedAnky) -> Int64 {
        writingDurationMs(parsed)
    }

    public static func isComplete(_ parsed: ParsedAnky) -> Bool {
        writingDurationMs(parsed) >= completeRitualMs
    }

    public static func formatted(_ durationMs: Int64) -> String {
        let totalSeconds = max(0, durationMs / 1000)
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return "\(minutes)m \(String(format: "%02d", seconds))s"
    }

    public static func clock(_ durationMs: Int64) -> String {
        let totalSeconds = max(0, durationMs / 1000)
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return "\(minutes):\(String(format: "%02d", seconds))"
    }

    public static func ritualClock(_ durationMs: Int64) -> String {
        clock(min(durationMs, completeRitualMs))
    }

    public static func utcDayProgress(at date: Date, secondsPerDay: TimeInterval = 86_400) -> Double {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        let start = calendar.startOfDay(for: date)
        let elapsed = date.timeIntervalSince(start)
        return min(1, max(0, elapsed / secondsPerDay))
    }
}

File: apps/android/app/src/main/java/inc/anky/android/core/protocol/AnkyDuration.kt
package inc.anky.android.core.protocol

import java.time.Instant
import java.time.ZoneOffset

object AnkyDuration {
    const val CompleteRitualMinutes: Int = 8
    const val CompleteRitualMs: Long = CompleteRitualMinutes * 60 * 1000L
    const val TerminalSilenceMs: Long = 8000

    fun durationMs(parsed: ParsedAnky): Long =
        parsed.events.sumOf { it.deltaMs } + (parsed.terminalSilenceMs ?: 0)

    fun isComplete(parsed: ParsedAnky): Boolean =
        durationMs(parsed) >= CompleteRitualMs && parsed.terminalSilenceMs == TerminalSilenceMs

    fun formatted(durationMs: Long): String {
        val totalSeconds = maxOf(0, durationMs / 1000)
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return "${minutes}m ${seconds.toString().padStart(2, '0')}s"
    }

    fun clock(durationMs: Long): String {
        val totalSeconds = maxOf(0, durationMs / 1000)
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return "$minutes:${seconds.toString().padStart(2, '0')}"
    }

    fun utcDayProgress(at: Instant, secondsPerDay: Long = 86_400): Double {
        val start = at.atZone(ZoneOffset.UTC).toLocalDate().atStartOfDay(ZoneOffset.UTC).toInstant()
        val elapsed = at.epochSecond - start.epochSecond
        return (elapsed.toDouble() / secondsPerDay.toDouble()).coerceIn(0.0, 1.0)
    }
}
